src/main/java/com/google/devtools/build/lib/remote/GoogleChannelConnectionFactory.java

```diff
@@ -94,12 +94,21 @@ public Single<ChannelConnectionWithServerCapabilities> create() {
     return Single.fromCallable(
             () -> channelFactory.newChannel(target, proxy, options, interceptors))
         .flatMap(
-            channel ->
-                RxFutures.toSingle(() -> getAndVerifyServerCapabilities(channel), directExecutor())
-                    .map(
-                        serverCapabilities ->
-                            new ChannelConnectionWithServerCapabilities(
-                                channel, serverCapabilities)));
+            channel -> {
+              var serverCapabilitiesSingle =
+                  RxFutures.toSingle(
+                      () -> getAndVerifyServerCapabilities(channel), directExecutor());
+
+              // Don't issue GetCapabilities calls if the requirement is NONE because the endpoint,
+              // e.g. Remote Asset API, might not implement the API. See #20342.
+              if (requirement == ServerCapabilitiesRequirement.NONE) {
+                return Single.just(
+                    new ChannelConnectionWithServerCapabilities(channel, serverCapabilitiesSingle));
```

Comment:
Is the logic reversed? This to me reads as "if the requirement is NONE, then call getAndVerifyServerCapabilities".

Reply:
No. `serverCapabilitiesSingle` is a lazy evaluation of `getAndVerifyServerCapabilities`.

If the requirement is NONE, we store it and call `getAndVerifyServerCapabilities` only when `ChannelConnectionWithServerCapabilities#getCapabilities` is called.

Otherwise, we evaluate the single now which calls `getAndVerifyServerCapabilities` immediately.